README
YubiKey C Library: libyubikey
=============================

This package make up the low-level C software development kit for the
Yubico https://www.yubico.com[YubiKey] authentication device.

Building
--------

After downloading and unpacking the package tarball, you build it as
follows.

  ./configure
  make check
  sudo make install

Another way to build the package is with a hand written short
makefile, invoke using 'make -f simple.mk check'.  This is usually not
recommended, and you will have to read simple.mk for further
information.

Building from version controlled sources
----------------------------------------

Warning!  This is only for developers and if you don't understand what
you are doing, you should download the release files instead.

You may check out the sources using Git with the following command:

  git clone https://github.com/Yubico/yubico-c.git

This will create a directory 'yubico-c'.  Enter the directory:

  cd yubico-c

Recent versions of autoconf, automake and libtool must be installed,
if you want to use the normal ./configure based approach.  AsciiDoc's
a2x tool is also required for building the man pages.  (See simple.mk
if you want a slimmed down non-autoconf build approach.)

Generate the build system:

  autoreconf --install

Then refer to the user sections above regarding building.

Command-line tools
------------------

The 'modhex' program converts data between modhex, normal hex, and
binary form.

The 'ykparse' program decrypts and parses one OTP given the OTP and
the AES key corresponding to that YubiKey.

The 'ykgenerate' program constructs and encrypts one OTP given AES key
and internal data values.

Example usage:

[source, sh]
-------
  $ ./modhex test
  ifhgieif
  $ ./modhex -d ifhgieif; echo
  test
  $ ./modhex -h b565716f
  nghgibhv
  $ ./modhex -h -d nghgibhv
  b565716f
  $ ./ykparse ecde18dbe76fbd0c33330f1c354871db dteffujedcflcindvdbrblehecuitvjkjevvehjd
  warning: overlong token, ignoring prefix: dteffuje
  Input:
    token: dcflcindvdbrblehecuitvjkjevvehjd
            20 4a 07 b2 f2 1c 1a 36 30 e7 df 89 83 ff 36 82
    aeskey: ecde18dbe76fbd0c33330f1c354871db
            ec de 18 db e7 6f bd 0c 33 33 0f 1c 35 48 71 db
  Output:
            87 92 eb fe 26 cc 13 00 a8 c0 00 10 b4 08 6f 5b

  Struct:
    uid: 87 92 eb fe 26 cc
    counter: 19 (0x0013)
    timestamp (low): 49320 (0xc0a8)
    timestamp (high): 0 (0x00)
    session use: 16 (0x10)
    random: 2228 (0x8b4)
    crc: 23407 (0x5b6f)

  Derived:
    cleaned counter: 19 (0x0013)
    modhex uid: jikdunvudhrr
    triggered by caps lock: no
    crc: F0B8
    crc check: ok
  $ ./ykparse ecde18dbe76fbd0c33330f1c354871db dteffujehknhfjbrjnlnldnhcujvddbikngjrtgh
  warning: overlong token, ignoring prefix: dteffuje
  Input:
    token: hknhfjbrjnlnldnhcujvddbikngjrtgh
            69 b6 48 1c 8b ab a2 b6 0e 8f 22 17 9b 58 cd 56
    aeskey: ecde18dbe76fbd0c33330f1c354871db
            ec de 18 db e7 6f bd 0c 33 33 0f 1c 35 48 71 db
  Output:
            87 92 eb fe 26 cc 13 00 30 c2 00 11 c8 9f 23 c8

  Struct:
    uid: 87 92 eb fe 26 cc
    counter: 19 (0x0013)
    timestamp (low): 49712 (0xc230)
    timestamp (high): 0 (0x00)
    session use: 17 (0x11)
    random: 40904 (0x9fc8)
    crc: 51235 (0xc823)

  Derived:
    cleaned counter: 19 (0x0013)
    modhex uid: jikdunvudhrr
    triggered by caps lock: no
    crc: F0B8
    crc check: ok
  $ ./ykgenerate ecde18dbe76fbd0c33330f1c354871db 8792ebfe26cc 0013 c0a8 00 10
  hcfktknicvbcnhbkvigcfhgddhhhrknc
  $ ./ykparse ecde18dbe76fbd0c33330f1c354871db `./ykgenerate ecde18dbe76fbd0c33330f1c354871db 8792ebfe26cc 0013 c0a8 00 10`
  Input:
    token: lihggvrgffbjnrehcgnkvknjkvubeekr
            a7 65 5f c5 44 18 bc 36 05 b9 f9 b8 9f e1 33 9c
    aeskey: ecde18dbe76fbd0c33330f1c354871db
            ec de 18 db e7 6f bd 0c 33 33 0f 1c 35 48 71 db
  Output:
            87 92 eb fe 26 cc 13 00 a8 c0 00 10 95 44 ec e9

  Struct:
    uid: 87 92 eb fe 26 cc
    counter: 19 (0x0013)
    timestamp (low): 49320 (0xc0a8)
    timestamp (high): 0 (0x00)
    session use: 16 (0x10)
    random: 17557 (0x4495)
    crc: 59884 (0xe9ec)

  Derived:
    cleaned counter: 19 (0x0013)
    modhex uid: jikdunvudhrr
    triggered by caps lock: no
    crc: F0B8
    crc check: ok
  $
-------

License
-------

The project is licensed under a BSD license.  See the file COPYING for
exact wording.  For any copyright year range specified as YYYY-ZZZZ in
this package note that the range specifies every single year in that
closed interval.


NEWS
Libyubikey NEWS -- History of user-visible changes.             -*- outline -*-

* Version 1.14 (unreleased)
 ** ykgenerate accepts optional random value to use.
 ** ykparse doesn't crash when only one parameter is given.
 ** Better self tests.
 ** Fix memory leak.
 ** Document modhex alphabet.
 ** Add different ModHex alphabets

* Version 1.13 (released 2015-03-05)
 ** Correct hex decode with uneven characters, for example "abc".

* Version 1.12 (released 2014-06-11)
 ** Rewrote man pages using Asciidoc.

* Version 1.11 (released 2013-10-18)
 ** Add man pages.
 ** Self-tests may be run under valgrind to catch bugs.
    Disable with "./configure --disable-valgrind-tests".

* Version 1.10 (released 2013-03-12)
 ** Added ./configure --enable-gcc-warnings to enable a lot of warnings.
 ** Warning fixes, build fixes etc.

* Version 1.9 (released 2012-05-31)
 ** Updated ld-version-script from gnulib to silence warnings.
 ** Fix out-of-tree builds.

* Version 1.8 (released 2012-03-13)
 ** modhex tool: Fix memory leak.
 ** Fix insecure rpath.

* Version 1.7 (released 2010-11-24)
 ** yubikey.h: Possible to use from C++ using extern namespace scoping.
    Reported by Giampaolo Bellini <iw2lsi@gmail.com>.

* Version 1.6 (released 2010-08-03)
 ** New API to generate OTPs.
    The API is yubikey_generate.  An helper function yubikey_aes_encrypt
    was added too.  Added YUBIKEY_OTP_SIZE constant for OTP buffer size.
 ** ykgenerate: New tool to generate OTPs.
 ** ykdebug/ykparse: The old tool "ykdebug" has been renamed to "ykparse".

* Version 1.5 (released 2008-09-31)
 ** The library now use a linker version script.
    It falls back to using Libtool's -export-symbols-regex if linker
    version scripts are not supported.
 ** Added new APIs to check validity of hex/modhex strings.
    The APIs are yubikey_modhex_p and yubikey_hex_p.

* Version 1.4 (released 2009-04-30)
 ** Improved instructions in README on how to build the package.
 ** Add interfaces yubikey_hex_encode and yubikey_hex_decode.
 ** The ykdebug tool now takes hex encoded AES keys instead of modhex encoded.

* Version 1.3 (released 2009-03-25)
 ** Add YUBIKEY_MODHEX_MAP.
 ** Use char* instead of uint8_t* for modhex functions.

* Version 1.2 (released 2009-01-13)
 ** Don't link to libpam.
 ** Add libtool -export-symbols-regex to restrict exported symbols.

* Version 1.1 (released 2008-06-10)
 ** Include simple.mk in distribution.

* Version 1.0 (released 2008-06-10)
 ** Uses autoconf, automake and libtool.
 ** Renamed function namespace from pof_* to yubikey_*.


ykaes.c
/* ykaes.c --- Implementation of AES-128.
 *
 * Copyright (c) 2006-2013 Yubico AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    * Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *
 *    * Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include "yubikey.h"

#define	NUMBER_OF_ROUNDS	10

static const uint8_t RC[] =
  { 0x1, 0x2, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80, 0x1B, 0x36 };

static const uint8_t rijndael_sbox[] = {
  0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5,
  0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
  0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0,
  0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
  0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC,
  0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
  0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A,
  0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
  0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0,
  0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
  0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B,
  0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
  0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85,
  0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
  0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5,
  0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
  0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17,
  0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
  0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88,
  0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
  0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C,
  0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
  0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9,
  0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
  0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6,
  0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
  0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E,
  0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
  0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94,
  0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
  0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68,
  0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16
};

static const uint8_t rijndael_inv_sbox[] = {
  0x52, 0x09, 0x6A, 0xD5, 0x30, 0x36, 0xA5, 0x38,
  0xBF, 0x40, 0xA3, 0x9E, 0x81, 0xF3, 0xD7, 0xFB,
  0x7C, 0xE3, 0x39, 0x82, 0x9B, 0x2F, 0xFF, 0x87,
  0x34, 0x8E, 0x43, 0x44, 0xC4, 0xDE, 0xE9, 0xCB,
  0x54, 0x7B, 0x94, 0x32, 0xA6, 0xC2, 0x23, 0x3D,
  0xEE, 0x4C, 0x95, 0x0B, 0x42, 0xFA, 0xC3, 0x4E,
  0x08, 0x2E, 0xA1, 0x66, 0x28, 0xD9, 0x24, 0xB2,
  0x76, 0x5B, 0xA2, 0x49, 0x6D, 0x8B, 0xD1, 0x25,
  0x72, 0xF8, 0xF6, 0x64, 0x86, 0x68, 0x98, 0x16,
  0xD4, 0xA4, 0x5C, 0xCC, 0x5D, 0x65, 0xB6, 0x92,
  0x6C, 0x70, 0x48, 0x50, 0xFD, 0xED, 0xB9, 0xDA,
  0x5E, 0x15, 0x46, 0x57, 0xA7, 0x8D, 0x9D, 0x84,
  0x90, 0xD8, 0xAB, 0x00, 0x8C, 0xBC, 0xD3, 0x0A,
  0xF7, 0xE4, 0x58, 0x05, 0xB8, 0xB3, 0x45, 0x06,
  0xD0, 0x2C, 0x1E, 0x8F, 0xCA, 0x3F, 0x0F, 0x02,
  0xC1, 0xAF, 0xBD, 0x03, 0x01, 0x13, 0x8A, 0x6B,
  0x3A, 0x91, 0x11, 0x41, 0x4F, 0x67, 0xDC, 0xEA,
  0x97, 0xF2, 0xCF, 0xCE, 0xF0, 0xB4, 0xE6, 0x73,
  0x96, 0xAC, 0x74, 0x22, 0xE7, 0xAD, 0x35, 0x85,
  0xE2, 0xF9, 0x37, 0xE8, 0x1C, 0x75, 0xDF, 0x6E,
  0x47, 0xF1, 0x1A, 0x71, 0x1D, 0x29, 0xC5, 0x89,
  0x6F, 0xB7, 0x62, 0x0E, 0xAA, 0x18, 0xBE, 0x1B,
  0xFC, 0x56, 0x3E, 0x4B, 0xC6, 0xD2, 0x79, 0x20,
  0x9A, 0xDB, 0xC0, 0xFE, 0x78, 0xCD, 0x5A, 0xF4,
  0x1F, 0xDD, 0xA8, 0x33, 0x88, 0x07, 0xC7, 0x31,
  0xB1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xEC, 0x5F,
  0x60, 0x51, 0x7F, 0xA9, 0x19, 0xB5, 0x4A, 0x0D,
  0x2D, 0xE5, 0x7A, 0x9F, 0x93, 0xC9, 0x9C, 0xEF,
  0xA0, 0xE0, 0x3B, 0x4D, 0xAE, 0x2A, 0xF5, 0xB0,
  0xC8, 0xEB, 0xBB, 0x3C, 0x83, 0x53, 0x99, 0x61,
  0x17, 0x2B, 0x04, 0x7E, 0xBA, 0x77, 0xD6, 0x26,
  0xE1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0C, 0x7D
};

static inline uint8_t
xtime (uint8_t b)
{
  return (b & 0x80) ? ((b << 1) ^ 0x1b) : (b << 1);
}

void
yubikey_aes_decrypt (uint8_t * state, const uint8_t * key)
{
  uint8_t i, j, round_key[0x10];
  uint8_t a02x, a13x;
  uint8_t a02xx, a13xx;
  uint8_t k1, k2;

  memcpy (round_key, key, sizeof (round_key));
  for (i = 0; i < NUMBER_OF_ROUNDS; i++)
    {
      round_key[0] ^= RC[i];

      round_key[0] ^= rijndael_sbox[round_key[13]];
      round_key[1] ^= rijndael_sbox[round_key[14]];
      round_key[2] ^= rijndael_sbox[round_key[15]];
      round_key[3] ^= rijndael_sbox[round_key[12]];

      for (j = 4; j < 16; j++)
	round_key[j] ^= round_key[j - 4];
    }
  for (i = 0; i < 0x10; i++)
    state[i] ^= round_key[i];

  for (i = 1; i <= NUMBER_OF_ROUNDS; i++)
    {
      /* First row: 0 shift, 0 4 8 12 */
      state[0] = rijndael_inv_sbox[state[0]];
      state[4] = rijndael_inv_sbox[state[4]];
      state[8] = rijndael_inv_sbox[state[8]];
      state[12] = rijndael_inv_sbox[state[12]];

      /* Second row: -1 shift, 1 5 9 13 */
      j = state[13];
      state[13] = rijndael_inv_sbox[state[9]];
      state[9] = rijndael_inv_sbox[state[5]];
      state[5] = rijndael_inv_sbox[state[1]];
      state[1] = rijndael_inv_sbox[j];

      /* Third row: -2 shift, 2 6 10 14 */
      j = state[2];
      state[2] = rijndael_inv_sbox[state[10]];
      state[10] = rijndael_inv_sbox[j];
      j = state[6];
      state[6] = rijndael_inv_sbox[state[14]];
      state[14] = rijndael_inv_sbox[j];

      /* Fourth row: -3 shift, 3 7 11 15 */
      j = state[3];
      state[3] = rijndael_inv_sbox[state[7]];
      state[7] = rijndael_inv_sbox[state[11]];
      state[11] = rijndael_inv_sbox[state[15]];
      state[15] = rijndael_inv_sbox[j];

      for (j = 15; j > 3; j--)
	round_key[j] ^= round_key[j - 4];

      round_key[0] ^=
	(RC[NUMBER_OF_ROUNDS - i] ^ rijndael_sbox[round_key[13]]);

      round_key[1] ^= rijndael_sbox[round_key[14]];
      round_key[2] ^= rijndael_sbox[round_key[15]];
      round_key[3] ^= rijndael_sbox[round_key[12]];

      for (j = 0; j < 16; j++)
	state[j] ^= round_key[j];
      if (i != NUMBER_OF_ROUNDS)
	{
	  for (j = 0; j < 16; j += 4)
	    {
	      k1 = state[j] ^ state[j + 2];
	      a02x = xtime (k1);
	      k2 = state[j + 1] ^ state[j + 3];
	      a13x = xtime (k2);

	      k1 ^= (k2 ^ xtime (state[j + 1] ^ state[j + 2]));
	      k2 = k1;

	      a02xx = xtime (a02x);
	      a13xx = xtime (a13x);

	      k1 ^= (xtime (a02xx ^ a13xx) ^ a02xx);
	      k2 ^= (xtime (a02xx ^ a13xx) ^ a13xx);

	      state[j] ^= (k1 ^ a02x);
	      state[j + 1] ^= k2;
	      state[j + 2] ^= (k1 ^ a13x);
	      state[j + 3] ^= (k2 ^ a02x ^ a13x);
	    }
	}

    }
}


/*******************************************************************
function aesEncrypt encrypts a single AES-128 block

  void aesEncrypt(unsigned char *state, const unsigned char *key)

Where:
     "state" is state buffer, i.e. plaintext in, ciphertext out
     "key" is pointer to AES key

*************************************************************************/

void
yubikey_aes_encrypt (uint8_t * state, const uint8_t * key)
{
  unsigned char i, j, k, tmp, round_key[0x10];

  memcpy (round_key, key, sizeof (round_key));

  for (i = 0; i < 16; i++)
    state[i] ^= key[i];

  for (i = 0; i < NUMBER_OF_ROUNDS; i++)
    {
      /* First row: 0 shift, 0 4 8 12 */
      state[0] = rijndael_sbox[state[0]];
      state[4] = rijndael_sbox[state[4]];
      state[8] = rijndael_sbox[state[8]];
      state[12] = rijndael_sbox[state[12]];

      /* Second row: 1 shift, 1 5 9 13 */
      tmp = state[1];
      state[1] = rijndael_sbox[state[5]];
      state[5] = rijndael_sbox[state[9]];
      state[9] = rijndael_sbox[state[13]];
      state[13] = rijndael_sbox[tmp];

      /* Third row: 2 shift, 2 6 10 14 */
      tmp = state[2];
      state[2] = rijndael_sbox[state[10]];
      state[10] = rijndael_sbox[tmp];
      tmp = state[6];
      state[6] = rijndael_sbox[state[14]];
      state[14] = rijndael_sbox[tmp];

      /* Fourth row: 3 shift, 3 7 11 15 */
      tmp = state[15];
      state[15] = rijndael_sbox[state[11]];
      state[11] = rijndael_sbox[state[7]];
      state[7] = rijndael_sbox[state[3]];
      state[3] = rijndael_sbox[tmp];

      if (i != (NUMBER_OF_ROUNDS - 1))
	{
	  for (k = 0; k < 16; k += 4)
	    {

	      j = state[k] ^ state[k + 1];
	      tmp = j ^ state[k + 2] ^ state[k + 3];

	      j = xtime (j);

	      state[k] ^= (j ^ tmp);

	      j = state[k + 1] ^ state[k + 2];
	      j = xtime (j);

	      state[k + 1] ^= (j ^ tmp);

	      j = state[k + 2] ^ state[k + 3];
	      j = xtime (j);

	      state[k + 2] ^= (j ^ tmp);
	      state[k + 3] = state[k] ^ state[k + 1] ^ state[k + 2] ^ tmp;
	    }
	}

      round_key[0] ^= RC[i];

      round_key[0] ^= rijndael_sbox[round_key[13]];
      round_key[1] ^= rijndael_sbox[round_key[14]];
      round_key[2] ^= rijndael_sbox[round_key[15]];
      round_key[3] ^= rijndael_sbox[round_key[12]];

      for (k = 4; k < 16; k++)
	round_key[k] ^= round_key[k - 4];

      for (j = 0; j < 16; j++)
	state[j] ^= round_key[j];
    }
}
